### PR TITLE
batchnorm specify channel axis and performance optimizations for batchnorm

### PR DESCRIPTION
**Pull request by cjolivier01**
@piiswrong 
@madjam 

1) Batch norm takes channel_axis parameter

2) Optmizations -- CPU batchnorm performance (without SIMD) approaches MKL performance

BatchNormV1Prop<cpu> 2D:  Timing [Forward] 1972.71 ms, avg: 3.94541 ms X 500 passes
BatchNormV1Prop<cpu> 2D:  Timing [Backward] 13659.1 ms, avg: 27.3182 ms X 500 passes

MKL BatchNormProp<cpu> 2D:  Timing [Forward] 99.892 ms, avg: 0.199784 ms X 500 passes
MKL BatchNormProp<cpu> 2D:  Timing [Backward] 121.685 ms, avg: 0.24337 ms X 500 passes

BatchNormProp<cpu> 2D:  Timing [Forward] 174.854 ms, avg: 0.319708 ms X 500 passes
BatchNormProp<cpu> 2D:  Timing [Backward] 166.815 ms, avg: 0.33363 ms X 500 passes

BatchNormV1Prop<gpu> 2D:  Timing [Forward] 5.058 ms, avg: 0.010116 ms X 500 passes
BatchNormV1Prop<gpu> 2D:  Timing [Backward] 14.812 ms, avg: 0.029624 ms X 500 passes

BatchNormProp<gpu> 2D:  Timing [Forward] 1.711 ms, avg: 0.003422 ms X 500 passes
BatchNormProp<gpu> 2D:  Timing [Backward] 1.752 ms, avg: 0.003504 ms X 500 passes

3) GPU version is faster than CUDNN

### CODE REVIEW

File: src/operator/batch_norm-inl.h [hunk 1]
```diff
@@ -32,13 +32,16 @@ enum BatchNormOpOutputs {kOut, kMean, kVar};  // req, out_data
...
+constexpr int DEFAULT_CHANNEL_AXIS = 1;
+
 /*! \brief Parameters for BatchNoram operator */
 struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
   float eps;
   float momentum;
   bool fix_gamma;
   bool use_global_stats;
   bool output_mean_var;
+  int channel_axis;
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: channel_axis -> axis

File: src/operator/batch_norm-inl.h [hunk 2]
```diff
@@ -32,13 +32,16 @@ enum BatchNormOpOutputs {kOut, kMean, kVar};  // req, out_data
 enum BatchNormOpAuxiliary {kMovingMean, kMovingVar};  // aux_states
 }  // namespace batchnorm
 
+constexpr int DEFAULT_CHANNEL_AXIS = 1;
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: This should go into batchnorm name space

File: src/operator/batch_norm-inl.h [hunk 3]
```diff
@@ -207,21 +212,26 @@ class BatchNormProp : public OperatorProperty {
     CHECK_EQ(in_shape->size(), 3U) << "Input:[data, gamma, beta]";
     const TShape &dshape = in_shape->at(0);
 
+    CHECK_GE(param_.channel_axis, -1) << "Invalid channel axis: " << param_.channel_axis;
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: use proper parsing: axis = axis + ndim()

File: src/operator/batch_norm-inl.h [hunk 4]
```diff
@@ -329,6 +339,128 @@ class BatchNormProp : public OperatorProperty {
...
+#define __bn_hostdevinl__  __host__ __device__ __forceinline__
+#define __bn_localinline__ __forceinline__
+#else
+#define __bn_hostonly__
+#define __bn_hostdevinl__  inline
+#define __bn_localinline__ inline
+#endif
+
+template<typename DType>
+class BNTensor3 {
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: Why define this instead of using mshadow::Tensor<xpu, 3>.

File: src/operator/batch_norm-inl.h [hunk 5]
```diff
@@ -329,6 +339,128 @@ class BatchNormProp : public OperatorProperty {
   BatchNormParam param_;
 };  // class BatchNormProp
 
+namespace batchnorm {
+
+#if defined(__CUDACC__)
+#define __bn_hostonly__        __host__
+#define __bn_hostdevinl__  __host__ __device__ __forceinline__
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: These macros already exists in mshadow.
for example this is MSHADOW_XINLINE

File: src/operator/batch_norm-inl.h (hunk 1 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: I don;t understand this comment

File: src/operator/batch_norm-inl.h (hunk 4 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: You told me not to use mshadow::Tensor, so I've never inspected it

File: src/operator/batch_norm-inl.h (hunk 1 above)
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: rename the argument to axis

File: src/operator/batch_norm-inl.h (hunk 4 above)
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: Did I? What was the reason? I don't remember

File: src/operator/batch_norm-inl.h (hunk 3 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: I don't understand this comment

File: src/operator/batch_norm-inl.h (hunk 5 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok

File: src/operator/batch_norm-inl.h (hunk 4 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: I also don't see how mshadow::Tensor<xpu, 3> knows about channel axis

Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: I think I was referring to don't use mshadow's template evaluation features. Like Tensor x; x = 1;
Only using the data structure is fine.

But you can leave it like this for now.

File: src/operator/batch_norm-inl.h (hunk 2 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok

File: src/operator/batch_norm-inl.h (hunk 3 above)
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: axis = -1 means first dim from right, axis=-2 means second dim from the right. You are only special casing -1 here. We need to support all

Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: oh, I see. I wasn't aware of this requirement.

File: src/operator/batch_norm-inl.h [hunk 6]
```diff
@@ -39,6 +43,7 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
   bool fix_gamma;
   bool use_global_stats;
   bool output_mean_var;
+  int channel_axis;
```
Comment (piiswrong) [inline src/operator/batch_norm-inl.h]: rename to axis

Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok

File: src/operator/batch_norm-inl.h (hunk 4 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: ok

File: src/operator/batch_norm-inl.h (hunk 1 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: done

**Comment by cjolivier01** [comment]
Unrelated problem:  libdc1394 error: Failed to initialize libdc1394

File: src/operator/batch_norm-inl.h (hunk 3 above)
Comment (cjolivier01) [inline src/operator/batch_norm-inl.h]: done
